File: backend/src/main/java/com/cinepulse/service/BookingService.java
package com.cinepulse.service;

import com.cinepulse.dto.BookingRequestDTO;
import com.cinepulse.dto.BookingResponseDTO;
import com.cinepulse.entity.*;
import com.cinepulse.exception.InvalidBookingException;
import com.cinepulse.exception.ResourceNotFoundException;
import com.cinepulse.exception.SeatUnavailableException;
import com.cinepulse.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class BookingService {

    private final BookingRepository bookingRepository;
    private final UserRepository userRepository;
    private final ShowRepository showRepository;
    private final SeatRepository seatRepository;
    private final BookingSeatRepository bookingSeatRepository;

    public BookingService(BookingRepository bookingRepository,
                          UserRepository userRepository,
                          ShowRepository showRepository,
                          SeatRepository seatRepository,
                          BookingSeatRepository bookingSeatRepository) {
        this.bookingRepository = bookingRepository;
        this.userRepository = userRepository;
        this.showRepository = showRepository;
        this.seatRepository = seatRepository;
        this.bookingSeatRepository = bookingSeatRepository;
    }

    @Transactional
    public BookingResponseDTO createBooking(BookingRequestDTO request) {
        // 1. Validate User
        User user = userRepository.findById(request.getUserId())
                .orElseThrow(() -> new ResourceNotFoundException("User not found with ID: " + request.getUserId()));

        // 2. Validate Show
        Show show = showRepository.findById(request.getShowId())
                .orElseThrow(() -> new ResourceNotFoundException("Show not found with ID: " + request.getShowId()));

        // 3. Validate Seats exist
        List<Long> requestedSeatIds = request.getSeatIds();
        if (requestedSeatIds == null || requestedSeatIds.isEmpty()) {
            throw new InvalidBookingException("Booking request must contain at least one seat.");
        }

        List<Seat> seats = seatRepository.findAllById(requestedSeatIds);
        if (seats.size() != requestedSeatIds.size()) {
            throw new InvalidBookingException("One or more requested seat IDs are invalid.");
        }

        // 4. Verify seats belong to the screen used by the selected show
        Long showScreenId = show.getScreen().getId();
        for (Seat seat : seats) {
            if (!seat.getScreen().getId().equals(showScreenId)) {
                throw new InvalidBookingException("Seat " + seat.getRowLabel() + seat.getSeatNumber() + " does not belong to screen for this show.");
            }
        }

        // 5. Verify seats are available for THIS SHOW
        Set<Long> currentlyBookedSeatIds = bookingSeatRepository.findBookedSeatIdsByShowId(show.getId());
        for (Seat seat : seats) {
            if (currentlyBookedSeatIds.contains(seat.getId())) {
                throw new SeatUnavailableException("Seat " + seat.getRowLabel() + seat.getSeatNumber() + " is already booked for this show.");
            }
        }

        // 6. Calculate total amount
        BigDecimal seatPrice = show.getPrice();
        BigDecimal totalAmount = seatPrice.multiply(BigDecimal.valueOf(seats.size()));

        // 7. Generate Booking Reference
        String bookingReference = "CP-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();

        // 8. Create Booking
        Booking booking = new Booking(user, show, bookingReference, "CONFIRMED", totalAmount);

        // 9. Add BookingSeats
        for (Seat seat : seats) {
            BookingSeat bookingSeat = new BookingSeat(booking, show, seat, seatPrice);
            booking.addBookingSeat(bookingSeat);
        }

        Booking savedBooking = bookingRepository.save(booking);

        // 10. Convert to Response DTO
        List<String> seatLabels = seats.stream()
                .map(s -> s.getRowLabel() + s.getSeatNumber())
                .collect(Collectors.toList());

        return new BookingResponseDTO(
                savedBooking.getId(),
                savedBooking.getBookingReference(),
                savedBooking.getStatus(),
                show.getMovie().getTitle(),
                show.getScreen().getTheater().getName(),
                show.getScreen().getName(),
                show.getShowDate(),
                show.getShowTime(),
                seatLabels,
                savedBooking.getTotalAmount(),
                savedBooking.getBookedAt() != null ? savedBooking.getBookedAt() : LocalDateTime.now()
        );
    }

    @Transactional(readOnly = true)
    public BookingResponseDTO getBookingById(Long id) {
        Booking booking = bookingRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Booking not found with ID: " + id));
        return convertToResponseDTO(booking);
    }

    @Transactional(readOnly = true)
    public BookingResponseDTO getBookingByReference(String reference) {
        Booking booking = bookingRepository.findByBookingReference(reference)
                .orElseThrow(() -> new ResourceNotFoundException("Booking not found with reference: " + reference));
        return convertToResponseDTO(booking);
    }

    private BookingResponseDTO convertToResponseDTO(Booking booking) {
        Show show = booking.getShow();
        List<String> seatLabels = booking.getBookingSeats().stream()
                .map(bs -> bs.getSeat().getRowLabel() + bs.getSeat().getSeatNumber())
                .collect(Collectors.toList());

        return new BookingResponseDTO(
                booking.getId(),
                booking.getBookingReference(),
                booking.getStatus(),
                show.getMovie().getTitle(),
                show.getScreen().getTheater().getName(),
                show.getScreen().getName(),
                show.getShowDate(),
                show.getShowTime(),
                seatLabels,
                booking.getTotalAmount(),
                booking.getBookedAt()
        );
    }
}

File: backend/app/services/booking_service.py
import uuid
import datetime
from decimal import Decimal
from typing import List
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from fastapi import HTTPException, status
from app.models import User, Show, Seat, Booking, BookingSeat
from app.schemas import BookingRequestDTO, BookingResponseDTO


def create_booking(db: Session, request: BookingRequestDTO) -> BookingResponseDTO:
    # 1. Validate User
    user = db.query(User).filter(User.id == request.user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"User not found with ID: {request.user_id}"
        )

    # 2. Validate Show
    show = db.query(Show).filter(Show.id == request.show_id).first()
    if not show:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Show not found with ID: {request.show_id}"
        )

    # 3. Validate Seats exist
    if not request.seat_ids:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Booking request must contain at least one seat."
        )

    seats = db.query(Seat).filter(Seat.id.in_(request.seat_ids)).all()
    if len(seats) != len(request.seat_ids):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="One or more requested seat IDs are invalid."
        )

    # 4. Verify seats belong to screen used by selected show
    show_screen_id = show.screen_id
    for seat in seats:
        if seat.screen_id != show_screen_id:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Seat {seat.row_label}{seat.seat_number} does not belong to screen for this show."
            )

    # 5. Verify seats are available for THIS SHOW
    booked_records = (
        db.query(BookingSeat.seat_id)
        .filter(BookingSeat.show_id == show.id, BookingSeat.seat_id.in_(request.seat_ids))
        .all()
    )
    if booked_records:
        already_booked_seat_ids = {r[0] for r in booked_records}
        conflict_seat = next(s for s in seats if s.id in already_booked_seat_ids)
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Seat {conflict_seat.row_label}{conflict_seat.seat_number} is already booked for this show."
        )

    # 6. Calculate total price
    seat_price = Decimal(str(show.price))
    total_amount = seat_price * len(seats)

    # 7. Generate Booking Reference
    ref_uuid = str(uuid.uuid4())[:8].upper()
    booking_reference = f"CP-{ref_uuid}"

    # 8. Create Booking
    new_booking = Booking(
        user_id=user.id,
        show_id=show.id,
        booking_reference=booking_reference,
        status="CONFIRMED",
        total_amount=total_amount,
        booked_at=datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None)
    )

    try:
        db.add(new_booking)
        db.flush()  # populate new_booking.id

        # 9. Add BookingSeats
        for seat in seats:
            booking_seat = BookingSeat(
                booking_id=new_booking.id,
                seat_id=seat.id,
                show_id=show.id,
                price=seat_price
            )
            db.add(booking_seat)

        db.commit()
        db.refresh(new_booking)
    except IntegrityError:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="One or more selected seats were booked concurrently by another request."
        )

    # Build response
    seat_labels = [f"{s.row_label}{s.seat_number}" for s in seats]
    return BookingResponseDTO(
        booking_id=new_booking.id,
        booking_reference=new_booking.booking_reference,
        status=new_booking.status,
        movie_title=show.movie.title,
        theater_name=show.screen.theater.name,
        screen_name=show.screen.name,
        show_date=show.show_date,
        show_time=show.show_time,
        seats=seat_labels,
        total_amount=new_booking.total_amount,
        booked_at=new_booking.booked_at
    )


def get_booking_by_id(db: Session, booking_id: int) -> BookingResponseDTO:
    booking = db.query(Booking).filter(Booking.id == booking_id).first()
    if not booking:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Booking not found with ID: {booking_id}"
        )
    return _convert_to_response_dto(booking)


def get_booking_by_reference(db: Session, reference: str) -> BookingResponseDTO:
    booking = db.query(Booking).filter(Booking.booking_reference == reference).first()
    if not booking:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Booking not found with reference: {reference}"
        )
    return _convert_to_response_dto(booking)


def _convert_to_response_dto(booking: Booking) -> BookingResponseDTO:
    show = booking.show
    seat_labels = [f"{bs.seat.row_label}{bs.seat.seat_number}" for bs in booking.booking_seats]
    return BookingResponseDTO(
        booking_id=booking.id,
        booking_reference=booking.booking_reference,
        status=booking.status,
        movie_title=show.movie.title,
        theater_name=show.screen.theater.name,
        screen_name=show.screen.name,
        show_date=show.show_date,
        show_time=show.show_time,
        seats=seat_labels,
        total_amount=booking.total_amount,
        booked_at=booking.booked_at
    )
